Fix invalid nested int guarding in broadcast_shapes()

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145957

Fixes #145874

This PR takes the approach of updating the logic determining whether multiple shapes broadcast together to handle nested ints specially.

Possible alternative approach: don't update `broadcast_shapes()` + indicate that e.g. `Ne(j0, 1)` should statically evaluate to False. I briefly tried this but it wasn't straightforward. Is it better?

jbschlosser committed:
Fix invalid nested int guarding in broadcast_shapes()

[ghstack-poisoned]

bobrenjc93 commented:
@jbschlosser I'm reviewing this now but in the meantime can you add a test case? Feel free to graft the repro from https://github.com/pytorch/pytorch/issues/145874

jbschlosser commented:
@bobrenjc93 Great thank you! will bring the test over now

bobrenjc93 commented on torch/functional.py:
```diff
@@ -127,12 +130,23 @@ def broadcast_shapes(*shapes):
-                    ):
-                        continue
+
+                    # NB: handle nested ints specially to avoid invalid guarding on Ne(j0, 1).
```
this is the first time I've seen the j0 notation. Is this a jagged tensor specific symbol? Is there a reason it's not defined in `torch/utils/_sympy`?

```
# Invariant: there must not be a prefix which is a prefix of another string,
# as this introduces ambiguity
prefix_str = {
    SymT.SIZE: "s",  # integer
    SymT.UNBACKED_INT: "u",  # integer
    # Prefix z here is chosen to avoid false aliasing in symbol_is_type test
    # DO NOT add a "z" type.  You also need to avoid conflicts on these
    # prefixes but this is somewhat easier to manage
    SymT.FLOAT: "zf",
    SymT.UNBACKED_FLOAT: "zuf",
    SymT.TMP: "tmp",
    SymT.PRECOMPUTED_SIZE: "ps",
    SymT.INDEX: "i",
    SymT.R0_INDEX: "r0_",
    SymT.R1_INDEX: "r1_",
    SymT.TEMPLATE_INDEX: "idx",
    SymT.XBLOCK: "x",
    SymT.YBLOCK: "y",
    SymT.ZBLOCK: "z",
    SymT.INDIRECT: "indirect",  # false aliasing?
    SymT.VIEW: "view",
    SymT.HALIDE: "h",
}
```

jbschlosser replied:
> this is the first time I've seen the j0 notation. Is this a jagged tensor specific symbol?

yes, `j0` refers to a so-called "nested int", which represents the jagged structure of a particular dimension. Nested tensors with `layout=torch.jagged` have shapes like `torch.Size([5, j0, 10])`.

> Is there a reason it's not defined in torch/utils/_sympy?

Good question - I'm not responsible for introducing nested ints originally but it does seem like it should be defined there. cc @soulitzer if there's a reason this wasn't done

soulitzer replied:
We do actually have a `singleton_int` defined in  `torch/utils/_sympy`. Historically, nested int was named singleton_int but as we added more nested-specific pieces, it was eventually renamed to NestedInt. For the purpose of sympy inference, however, the logic remained generic, so the name wasn't changed here.

jbschlosser committed:
Update on "Fix invalid nested int guarding in broadcast_shapes()"


Fixes #145874

This PR takes the approach of updating the logic determining whether multiple shapes broadcast together to handle nested ints specially.

Possible alternative approach: don't update `broadcast_shapes()` + indicate that e.g. `Ne(j0, 1)` should statically evaluate to False. I briefly tried this but it wasn't straightforward. Is it better?

[ghstack-poisoned]

soulitzer commented:
As I'm working on enabling built-in ops like binary add to not fail when I have j0, j1 that have offsets with the same underlying data but are different Tensor object instances - I'm actually running into this same issue in a couple more places: (1) during backward I need to check whether next output is expandable to gradient, (2) during backward I need to reduce the gradient, if the is_expandable check succeeded, (3) the raggedness check we use for all nested ops during forward compare with -1. 

One thing we could do is special case for nested int in those comparisons, e.g. replace `x == 1` with `not is_nested_int(x) and x == 1`, but it might not be quite enough, e.g. for the case where I want to figure out whether j0 == j1, where I cannot simply bypass it. 

One workaround is introducing some kind of comparison that doesn't trigger any guard.
```python
def compare_without_guard(lhs, rhs):
    if isinstance(lhs.node, NestedIntNode):
        # No special logic needed in the eager case
        return lhs == rhs

    # Directly do comparison on the hint to avoid a guard if we are tracing
    lhs_hint = lhs.node.hint
    rhs_hint = rhs.node.hint
    return lhs_hint == rhs_hint
```
Maybe its worth thinking about these issues jointly?

bobrenjc93 commented:
@soulitzer @jbschlosser would you be open to a meeting w/ @laithsakka and I to discuss NJT/dynamic shapes interactions? We just posted our execution plan, but admittedly I was oblivious to the NJT specific workstream when writing it: https://fb.workplace.com/groups/8940092306109185/posts/8940111256107290

Given that NJT probably exercises the DS system the most and may influence our design decisions on things like dynamic int, it might make sense to sync earlier rather than later?

jbschlosser commented:
@bobrenjc93 yep sounds great!

bobrenjc93 committed:
Update on "Fix invalid nested int guarding in broadcast_shapes()"


Fixes #145874

This PR takes the approach of updating the logic determining whether multiple shapes broadcast together to handle nested ints specially.

Possible alternative approach: don't update `broadcast_shapes()` + indicate that e.g. `Ne(j0, 1)` should statically evaluate to False. I briefly tried this but it wasn't straightforward. Is it better?

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot rebase

bobrenjc93 committed:
Update

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
